multiple guardian CI testing

justinschuldt opened the pull request:
This change makes the tilt arg `--num=x` initialize devnet with the specified number of guardians. Now guardian set `0` will include however many guardians were specified.

The basis of this is a script, `./scripts/guardian-set-init.sh`, that creates the list of signers, and the governance messages for registering all the various contracts. The initialization data is written to `.env` files, which are copied to `./ethereum`, `./solana`, and `./terra`. The devnet bootstrapping code for each of the chains now reads from `.env` files, rather than using hardcoded governance VAA strings.

This change includes running the `tilt ci` test suite in GitHub actions with `--num=2`, which means that all of the integration tests in `./sdk/js` now also test the code paths for generating governance messages, p2p bootstraping, calculating quorum, validating signatures, etc.

I added a json file with all of the constants needed for generating the signer data and governance messages. I am hoping we can develop some patterns for creating and fetching that data, move away from duplicated strings.

I added a little documentation and tooling to `./clients/token_bridge/` and `./clients/nft_bridge/` to alleviate some of the pain-points I found trying to use them to generate governance VAAs with multiple guardians. Just basic "how do you run this" documentation and low hanging fruit in the typescript setup.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/certusone/wormhole/1047)
<!-- Reviewable:end -->

SEJeff commented on solana/devnet_setup.sh:
```diff
@@ -30,10 +30,14 @@ cli_address=6sbzC1eH4FTujJXWj51eQe25cYvr4xfXbJ1vAj7j2k5J
+# load the .env file with the devent init data
+source .env
+# remove brackets and double quotes, transform json list of strings to CSV with no quotes.
+guardians=$(echo $INIT_SIGNERS | sed -r 's/(\[|\]|\")//g')
```
Dirty. I love it.

justinschuldt replied:
replace this with a cleaner method - just passed the CSV string in from the .env file. Didn't love the dirty.

SEJeff commented on solana/devnet_setup.sh:
```diff
@@ -83,24 +87,25 @@ token-bridge-client create-meta "$nft" "Not a PUNK 2🎸" "PUNK2🎸" "https://w
-node main.js solana execute_governance_vaa $(node main.js generate_register_chain_vaa 4 0x0000000000000000000000000290FB167208Af455bB137780163b7B7a9a10C16)
+node main.js solana execute_governance_vaa "$REGISTER_ETH_TOKEN_BRIDGE_VAA"
+node main.js solana execute_governance_vaa "$REGISTER_TERRA_TOKEN_BRIDGE_VAA"
+node main.js solana execute_governance_vaa "$REGISTER_BSC_TOKEN_BRIDGE_VAA"
```
At the top where this has `set -x` strongly consider `set -xu` The `-u` will error out if any of these env vars are unset, which would just cause future sadness in the tests.

justinschuldt replied:
Will do, great suggestion!

SEJeff reviewed (approved):
What a beast of a PR, nice work!

You'll need to ensure `jq` is installed in the container the runs the setup and guardian init scripts, but otherwise, this is great stuff. I did leave some nits, but nothing insane that should block merging (once you fix the `jq` error).

justinschuldt commented on scripts/guardian-set-init.sh:
```diff
@@ -0,0 +1,112 @@
+echo "number of guardians to initialize: ${numGuardians}"
+
+addressesJson="./scripts/devnet-consts.json"
+
```
Perf, adding this, thanks.

justinschuldt commented on clients/nft_bridge/main.ts:
```diff
@@ -123,17 +126,17 @@ yargs(hideBin(process.argv))
-                describe: 'Guardian\'s secret key',
+                describe: 'Guardian\'s secret key, CSV allowed',
                 type: "string",
                 default: "cfb12303a19cde580bb4dd771639b0d26bc68353645571a8cff516ab2ee113a0"
             })
-    }, async (argv: any) => {
+    }, async (argv) => {
```
The `: any` here was overriding the actual, specific type annotations of `argv` with the `any` type. Removing the `: any` gets us more type hinting, intellisense, etc, just by not throwing away the native types of `argv` defined by the package that exports it.

justinschuldt replied:
no types with `argv: any`:
![argv-any](https://user-images.githubusercontent.com/2636098/162535294-f706d38c-7d10-41b8-9b80-817abe56be29.png)

types:
![argv-type](https://user-images.githubusercontent.com/2636098/162535422-b9997d12-6fe9-4d57-b845-95220e32dcc8.png)


noticed we were making it harder on ourselves by throwing away/not using the types we get for free.

justinschuldt commented on clients/nft_bridge/main.ts:
```diff
@@ -1,6 +1,6 @@
 import yargs from "yargs";
 
-const { hideBin } = require('yargs/helpers')
+import { hideBin } from 'yargs/helpers'
```
So tl;dr, the javascript module system is like 10 competing standards, and it is a mess.  There are a bunch of different ways of packaging and resolving modules (commonjs, esModules, etc), each with their own way of importing, doing introspection, etc.

The way you bring a module into your javascript project, require vs import, determines what code is actually pulled in. Due to the state of modules being a mess, most popular packages will export the same source code as multiple module types, to make it easy for users to consume it.

In this case, if you `const { hideBin } = require('yargs/helpers')`, you will only resolve the commonjs module, which has no type definitions. However, if you `import { hideBin } from 'yargs/helpers'`, then you will also resolve `@types/yargs`, which will supply typings for the library, allow the typescript compiler to do tree-shaking and lazy-loading (if applicable), and other general goodness.

For Typescript projects, if you can `import` it rather than `require` it, you pretty much always should. If you'd like to read about it, it's a hot mess all the way down, and it gets worse when you try to support both the browser and nodejs runtimes. [https://www.typescriptlang.org/docs/handbook/module-resolution.html](https://www.typescriptlang.org/docs/handbook/module-resolution.html)

justinschuldt replied:
types from require:
![require-ts](https://user-images.githubusercontent.com/2636098/162538342-29a6e86f-98ed-48dc-b178-5f70c4688245.png)

types from import:
![import-ts](https://user-images.githubusercontent.com/2636098/162538385-260bdec5-f2d8-40cd-adab-139cfe529d07.png)

This is low hanging fruit - don't throw away types, use what is supplied by the module.

justinschuldt commented on scripts/guardian-set-init.sh:
```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env bash
+# This script allows devnet initalization with more than one guardian.
+# First argument is the number of guardians for the initial guardian set.
+set -e
```
Great suggestion, that's much better. Added it, thanks!

justinschuldt commented on scripts/devnet-consts.json:
```diff
@@ -0,0 +1,351 @@
+                "testNFT": {
+                    "address": "0x5b9b42d6e4B2e4Bf8d42Eba32D46918e10899B66",
+                    "name": "Not an APE 🐒",
+                    "symbol": "APE🐒",
```
[Monkey see](https://github.com/certusone/wormhole/blob/dev.v2/ethereum/scripts/deploy_test_token.js#L67-L69), Monkey ctrl+c, ctrl+v

justinschuldt commented on clients/token_bridge/main.ts:
```diff
@@ -1,6 +1,6 @@
 import yargs from "yargs";
 
-const {hideBin} = require('yargs/helpers')
+import { hideBin } from 'yargs/helpers'
```
same as the other convo: https://github.com/certusone/wormhole/pull/1047/files#r841847226

justinschuldt commented:
> You'll need to ensure `jq` is installed in the container the runs the setup and guardian init scripts, but otherwise, this is great stuff. I did leave some nits, but nothing insane that should block merging (once you fix the `jq` error).

@SEJeff The latest commit moves the `guardian-set-init.sh` into a docker container in order to supply the dependencies (`jq`), and just be cleaner generally. Functionality is the same, just added a Dockerfile which executes the bash script. I incorporated your suggestions, all good stuff, appreciate the feedback.

justinschuldt reviewed (commented):
> Reviewed 23 of 28 files at r1, 4 of 5 files at r2, 3 of 3 files at r3, all commit messages.
> Dismissed @SEJeff from 8 discussions.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/1047)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @leoluk)



<!-- Sent from Reviewable.io -->

SEJeff commented on scripts/guardian-set-init.sh:
```diff
@@ -45,68 +80,89 @@ solNFTBridge=$(jq --raw-output '.chains."1".contracts.nftBridgeEmitterAddress' $
 
+# 7) copy the local .env file to the solana & terra dirs, if the script is running on the host machine
+# chain dirs will not exist if running in docker for Tilt, only if running locally. check before copying.
+if [[ -d ./ethereum ]]; then
+    echo "copying $ethFile to /etherum/.env"
+    cp $ethFile ./ethereum/.env
+fi
+if [[ -d ./solana ]]; then
+    echo "copying $envFile to /solana.env"
+    cp $envFile ./solana/.env
+fi
+if [[ -d ./terra/tools ]]; then
+    echo "copying $envFile to /terra/tools/.env"
+    cp $envFile ./terra/tools/.env
+fi
```
```
for envfile in /ethereum/.env /solana/.env /terra/tools/.env; do
    dirname=$(echo $envfile | dirname)
    if [ -d "$dirname" ]; then
        echo "copying $envFile to $envfile"
        cp $envFile ./$envfile
    fi
done
```

Maybe more magical than you'd like, but this is less code (untested, just typed it in like a monkey).

If you'd rather keep what you have, fix the comment for solana to have the correct path: `/solana.env` --> `/solana/.env`

justinschuldt replied:
Implemented, thanks.

SEJeff reviewed (dismissed):
Great stuff!

claudijd commented on docs/devnet.md:
```diff
@@ -29,56 +29,6 @@
 The terra testnet can be used just like a normal localterra network (can be selected in finder and station).
 
-### Ganche default accounts
+### devent guardian addresses
```
Typo: devent

justinschuldt replied:
fixed, good catch, thanks.

justinschuldt commented:
*[docs/devnet.md, line 32 at r3](https://reviewable.io/reviews/certusone/wormhole/1047#-MzaSe7lTylLKrRtYqcR-r3-32:-N-PbdH6EDCunVlb2kZS:b8gj3mz) ([raw file](https://github.com/certusone/wormhole/blob/6b82363872854e9a2f5bd0bf77704e649cc9033e/docs/devnet.md#L32)):*
<details><summary><i>Previously, claudijd (Jonathan Claudius) wrote…</i></summary><blockquote>

Typo: devent
</blockquote></details>

Good catch, fixed.


<!-- Sent from Reviewable.io -->

justinschuldt commented:
*[scripts/guardian-set-init.sh, line 166 at r2](https://reviewable.io/reviews/certusone/wormhole/1047#-MzaSe7moKk7fCRY3ZNh-r2-166:-N-PbmsT6G8XBRVZtk_L:b-fznogt) ([raw file](https://github.com/certusone/wormhole/blob/23271c205c06bcacf5fafbc557be07f782edef67/scripts/guardian-set-init.sh#L166)):*
<details><summary><i>Previously, SEJeff (Jeff Schroeder) wrote…</i></summary><blockquote>

```
for envfile in /ethereum/.env /solana/.env /terra/tools/.env; do
    dirname=$(echo $envfile | dirname)
    if [ -d "$dirname" ]; then
        echo "copying $envFile to $envfile"
        cp $envFile ./$envfile
    fi
done
```

Maybe more magical than you'd like, but this is less code (untested, just typed it in like a monkey).

If you'd rather keep what you have, fix the comment for solana to have the correct path: `/solana.env` --> `/solana/.env`
</blockquote></details>

Good idea, done.


<!-- Sent from Reviewable.io -->

justinschuldt reviewed (commented):
> Reviewed 3 of 3 files at r4, all commit messages.
> Dismissed @SEJeff and @claudijd from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/certusone/wormhole/1047)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @leoluk)



<!-- Sent from Reviewable.io -->

justinschuldt commented:
rebased on `dev.v2` because GH's "rebase+merge" was not able to.
